client/src/sync-filesystem.js

```diff
@@ -6,39 +6,216 @@
+    });
+
+    fsUtils.getPathsToSync(fs, root, function(err, pathsToSync) {
+      if(err) {
+        return callback(err);
+      }
+
+      var toSync = pathsToSync.toSync;
+
+      // Ignore redundancies
+      paths.forEach(function(path) {
+        if(toSync.indexOf(path) === -1) {
+          pathsToSync.toSync.push(path);
+        }
+      });
+
+      fsUtils.setPathsToSync(fs, root, pathsToSync, callback);
+    });
+  };
+
+  // Get the path that was modified during a sync
+  self.getModifiedPath = function(callback) {
+    fsUtils.getPathsToSync(fs, root, function(err, pathsToSync) {
+      if(err) {
+        return callback(err);
+      }
+
+      callback(null, pathsToSync && pathsToSync.modified);
+    });
+  };
+
+  // Indicate that the path on top of the queue has
+  // begun syncing
+  self.setSyncing = function(callback) {
+    fsUtils.getPathsToSync(fs, root, function(err, pathsToSync) {
+      if(err) {
+        return callback(err);
+      }
+
+      if(!pathsToSync || !pathsToSync.toSync || !pathsToSync.toSync[0]) {
+        return callback();
+      }
+
+      pathsToSync.toSync[0].syncing = true;
+
+      callback();
+    });
+  };
+
+  // Delay the sync of the currently syncing path
+  // by moving it to the end of the sync queue
+  self.delaySync = function(callback) {
+    fsUtils.getPathsToSync(fs, root, function(err, pathsToSync) {
+      if(err) {
+        return callback(err);
+      }
+
+      if(!pathsToSync || !pathsToSync.toSync || !pathsToSync.toSync[0]) {
+        return callback();
+      }
+
+      var delayedSync = pathsToSync.toSync.shift();
+      pathsToSync.toSync.push(delayedSync);
+      delete pathsToSync.modified;
+
+      fsUtils.setPathsToSync(fs, root, pathsToSync, function(err) {
+        if(err) {
+          return callback(err);
+        }
+
+        callback(null, delayedSync.path);
+      });
+    });
+  };
+
+  // Remove the path that was just synced
+  self.dequeueSync = function(callback) {
+    fsUtils.getPathsToSync(fs, root, function(err, pathsToSync) {
+      if(err) {
+        return callback(err);
+      }
+
+      if(!pathsToSync || !pathsToSync.toSync || !pathsToSync.toSync[0]) {
+        return callback();
+      }
+
+      var removedSync = pathsToSync.toSync.shift();
+      if(!pathsToSync.toSync.length) {
+        delete pathsToSync.toSync;
+      }
+      delete pathsToSync.modified;
+
+      fsUtils.setPathsToSync(fs, root, pathsToSync, function(err) {
+        if(err) {
+          return callback(err);
+        }
+
+        callback(null, pathsToSync.toSync, removedSync.path);
+      });
+    });
+  };
+
+  // Set the sync root for the filesystem.
+  // The path provided must name an existing directory
+  // or the setter will fail.
+  // Once the new root is set, the paths remaining to
+  // sync and the path that was modified during a sync
+  // are filtered out if they are not under the new root.
+  self.setRoot = function(newRoot, callback) {
```

Comment:
I don't know if this would make this simpler or more difficult, but we could consider forcing the root to be a constant for the lifetime of the client instance (i.e., you'd have to create a new client if you wanted to change it later).